Review

Ashwing43 opened the pull request.

Ashwing43 committed:
Initial commit

Ashwing43 committed:
Added feature to cancel order

Ashwing43 committed:
Added logic for displaying un-cancelled orders to customer and admin

Ashwing43 committed:
Added logic for displaying place order in GetOrderStatus function.

Ashwing43 committed:
Added logic for displaying place order in GetOrderStatus function.

Ashwing43 committed:
Added feature for 'back' button

Ashwing43 committed:
Added functionality for backing from login page

Ashwing43 committed:
Refactored code

Ashwing43 commented:
Note: Type back in any prompt to go back to previous menu.

vitthal-balkawade-coditas commented on .gitignore:
We usually don't push this file to code review as it is intended for your personal workspace settings.

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+            _userService = userService;
+            _productService = productService;
+
+            OrderService orderService1 = (OrderService)orderService;
```
orderService1 is not a standard naming convention. You should always use meaningful names.

Why do you need another instance of Order service here. You already have an instance, can you not make use of it?

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+        {
+            try
+            {
+                Customer user = (Customer)_userService.GetUserById(userId);
```
Why do you need to explicitly type cast to Customer? Is 'GetUserById' not returning Customer type?

Ashwing43 replied:
'GetUserById' returns an object of type 'User' which is a abstract class and parent of 'Customer' and 'Admin' class. but in this function I want to access a 'List' of Addresses defined in the class 'Customer' that's why I had to type cast this object to 'Customer' class.
How should do this then sir?
Is there any better approach?
Thank you.

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+            orderService1.OnOrderProcessed += HandleOrderProcessed;
+        }
+
+        public void PlaceOrder(Guid userId)
```
Here 'PlaceOrder' function should only be responsible for placing an order i.e. single responsibility.

All other logic should have their own methods.

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+                    while (string.IsNullOrWhiteSpace(choice))
+                    {
+                        choice = Console.ReadLine();
+                        if(choice == "back") {
```
Never hard code constant values, here you should move value 'back' to readonly constants.

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+                        {
+                            Console.WriteLine("Input cannot be empty.");
+                        }
+                        else if (choice == "1" || choice == "2")
```
No magic numbers, move it to constants

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+                    Console.WriteLine("2.Finish order");
+
+                    var choice = "";
+                    while (string.IsNullOrWhiteSpace(choice))
```
Move this while loop to a new function to have separate responsibility and better abstraction.

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+            }
+        }
+
+        public void ShowAllOrders(Guid UserId)
```
ShowAllOrdersByUserId

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+            }
+        }
+
+        public void GetOrderStatus(Guid UserId)
```
Function name says get something, however, you are not returning anything

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+                try
+                {
+                    Order order = _orderService.GetOrderById(id);
+                    if (order is null)
```
Do not leave blocks without braces even if it is a single line block.

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+                        Console.Clear();
+                        return;
+                    }
+                    if (string.IsNullOrWhiteSpace(input))
+                    {
+                        Console.WriteLine("Input cannot be empty.");
+                    }
+                    else if (Guid.TryParse(input, out Guid k))
+                    {
+                        id = k;
+                        break;
+                    }
```
Try to avoid IF-ELSE-IF code structure.

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,379 @@
+            }
+        }
+
+        private void HandleOrderProcessed(Order o, string statused)
```
statused??

vitthal-balkawade-coditas replied:
May be DisplayOrderStatus or ShowOrderStatus?

vitthal-balkawade-coditas commented on Controller/ProductController.cs:
```diff
@@ -0,0 +1,349 @@
+                        Console.Clear();
+                        return;
+                    }
+                    if (string.IsNullOrWhiteSpace(quantityInput))
+                    {
+                        Console.WriteLine("Quantity cannot be empty, Please try again.");
+                    }
```
Convert this into a guard clause so that you can avoid this if-else-if

vitthal-balkawade-coditas commented on Controller/ProductController.cs:
```diff
@@ -0,0 +1,349 @@
+            }
+        }
+
+        public void ShowAllProductsCustomer()
```
ShowAllProductsCustomer??

vitthal-balkawade-coditas commented on Controller/ProductController.cs:
```diff
@@ -0,0 +1,349 @@
+            }
+        }
+
+        public void ShowAllProductsAdmin()
```
This method has common code with method 'ShowAllProductsCustomer', can you not extract common code and avoid duplication?

vitthal-balkawade-coditas replied:
ShowAllProductsAdmin??

vitthal-balkawade-coditas commented on Controller/ProductController.cs:
```diff
@@ -0,0 +1,349 @@
+            }
+        }
+
+        public void UpdateProductDetails()
```
This is a huge function, try to refactor it and have small functions handling single responsibility.

vitthal-balkawade-coditas commented on Controller/ProductController.cs:
```diff
@@ -0,0 +1,349 @@
+                    {
+                        Console.WriteLine("Input cannot be empty.");
+                    }
+                    else if (int.TryParse(choice, out int a))
```
Avoid using letters, numbers for vars, always use meaningful names.

vitthal-balkawade-coditas commented on Controller/ProductController.cs:
```diff
@@ -0,0 +1,349 @@
+                    }
+                }
+                Product prod = _productService.GetProductById(products[input - 1].Id);
+
```
Remove extra lines/spaces

vitthal-balkawade-coditas commented on CustomeLogger/log.txt:
Do not push this file to git

vitthal-balkawade-coditas reviewed (commented):
Good work Ashwin!

Have given you the initial feedback which mainly talks about
clean code, naming conventions etc.

Try to refactor code to have clean code so that it can be more readable.
Use meaningful names.

Please go through all review comments.

Ashwing43 committed:
Deleted unnecessary files.

Ashwing43 committed:
Moved delegates to new folder instead of defining them in OrderService.cs file

Updated IOrderService.cs that way we wont need to typecast them in OrderController.cs

Ashwing43 committed:
Renamed method to have meaningful names

Changed GetOrderStatus method to DisplayOrderStatus

Ashwing43 committed:
Moved hardcoded string 'back' to Constants.cs file and used its reference everywhere

Ashwing43 committed:
next commit

Ashwing43 committed:
next commit

Ashwing43 committed:
next commit

Ashwing43 committed:
Added feature to save data in json file

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,481 @@
+					}
+
+
+					if (choice == Choice_Constants.ONE)
```
Insetad 'One' you can make it more readable, e.g. AddProduct.
You can also use enums for this

vitthal-balkawade-coditas commented on Controller/OrderController.cs:
```diff
@@ -0,0 +1,481 @@
+			{
+				var orders = _orderService.GetAllOrdersByUserId(userId).ToList();
+				DisplayAllOrder(orders);
+				Loader.Loader.PressAnyKeyToExit();
```
Loader.Loader ??

vitthal-balkawade-coditas commented on Controller/ProductController.cs:
```diff
@@ -0,0 +1,361 @@
+					return;
+				}
+				
+				if (choice == Choice_Constants.ONE)
+				{
+					Console.WriteLine("Enter quantity to increase");
+					(int quantity, wantToGoBack) = GetQuantityInput();
+					if (wantToGoBack)
+					{
+						Console.Clear();
+						return;
+					}
+					
+					prod.Quantity += quantity;
+					_productService.Save();
+					Console.WriteLine("Quantity increased by " + quantity);
+					Loader.Loader.PressAnyKeyToExit();
+					return;
```
Move this logic to a new function for better responsibility and abstraction.

vitthal-balkawade-coditas commented on Controller/ProductController.cs:
```diff
@@ -0,0 +1,361 @@
+					return;
+				}
+
+				if (choice == "3")
```
Avoid magic numbers

vitthal-balkawade-coditas reviewed (commented):
Try to refactor big functions to small ones.

Format your code to have proper indentation. Check your editor format settings.

Ashwing43 committed:
next commit
